src/main/java/BasicMAPF/Solvers/PIBT/PIBT_Solver.java

```diff
@@ -0,0 +1,424 @@
+                return "invalid";
+            };
+        }
+        return "invalid";
+    }
+
+    /**
+     * helper function
+     * update priority of each agent in current timestamp using instance
+     */
+    private void updatePriorities(MAPF_Instance instance) {
+        for (Agent agent : this.priorities.keySet()) {
+            // agent reach his target
+            if (this.agentPlans.get(agent).containsTarget()) {
+                this.priorities.put(agent, -1.0);
+            }
+            else {
+                double currentPriority = this.priorities.get(agent);
+                this.priorities.put(agent, currentPriority + 1);
+            }
+        }
+    }
+
+    /**
+     * helper function to find all neighbors of single agent
+     * @param location to find his neighbors
+     * @return List contains all neighbors of current I_Location
+     */
+    private List<I_Location> findAllNeighbors(I_Location location) {
+        return location.outgoingEdges();
+    }
+
+    /**
+     * helper function to find best coordinate among candidates
+     * best will be the node which is closest to current agent goal
+     * distances are calculated using this.heuristic
+     * @param candidates list of candidates
+     * @param current Agent currently making decision
+     * @return I_Location - node who is closets to current goal among candidates
+     */
+    private I_Location findBest(List<I_Location> candidates, Agent current) {
+        I_Location bestCandidate = null;
+        Float minDistance = Float.MAX_VALUE;
+        for (I_Location location : candidates) {
+            Float distance = this.heuristic.getHToTargetFromLocation(current.target, location);
+            if (distance < minDistance) {
+                minDistance = distance;
+                bestCandidate = location;
+            }
+        }
+        return bestCandidate;
+    }
+
+    /**
+     * init priority of each agent in the beginning of the algorithm
+     * each agent have a unique double representing his priority
+     * update this.priorities
+     */
+    private void initPriority(MAPF_Instance instance) {
+        int numberOfAgents = instance.agents.size();
+        double uniqueFactor = 1.0 / numberOfAgents;
+        int i = 1;
+        for (Agent agent : instance.agents) {
+            // (uniqueFactor * i) is a unique representation for the priority of each agent
+            this.priorities.put(agent, uniqueFactor * i);
+            i++;
+        }
+    }
+
+    /**
+     * function to get map entry with maximum value
+     * extract agent with max priority
+     * @param prioritiesMap representing agents and their priority
+     * @return map entry <agent, priority>
+     */
+    public Map.Entry<Agent, Double> getMaxEntry(Map<Agent, Double> prioritiesMap) {
+        Map.Entry<Agent, Double> maxEntry = null;
+        for (Map.Entry<Agent, Double> entry : prioritiesMap.entrySet()) {
```

Comment:
TreeMap sorts by key - in this case we need the values to be sorted

Reply:
Yeah you're completely right. It can probably good to keep another data structure like a priority queue to make this quicker. But no need to do it for now.